[MLIR] Let matchers work on int ranges

gysit opened the pull request:
This commit adds three matchers that unlike the `m_NonZero` matcher not only match constants, but also operations that implement the InferIntRangeInterface. These matchers can then match a non-zero value or a value that is not minus one based on the inferred range. Additionally, the commit uses the new matchers in the `getSpeculatability` functions of Arith's signed and unsigned integer divisions. At the moment, the matchers only look at the defining operation to avoid expensive IR walks.

This range based matchers can be useful when hoisting divisions out of a loop, which requires knowing the divisor is non-zero and not minus one for signed divisions. Just checking for a constant divisor may not be sufficient, if the divisor is, for example, the result of an operation that returns the number of threads of a team of threads.

gysit committed:
[MLIR] Use InferIntRangeInterface in matchers

This commit adds a m_NonZeroUnsigned matcher that unlike the
m_NonZero matcher not only matches constants, but also operations that
implement the InferIntRangeInterface. It can then match a non-zero value
based on the inferred range. Additionally, the commit uses the new
matcher in the getSpeculatability function of Arith's unsigned integer
division. At the moment, the matcher only looks at the defining
operation to avoid expensive IR walks.

This range based matchers can be useful when hoisting divisions out of
a loop, which requires knowing the divisor is non-zero. Just checking
for a constant divisor may not be sufficient, if the divisor is,
for example, the result of an operation that returns the number of
threads of a team of threads.

Dinistro commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -100,6 +101,41 @@ struct constant_op_binder {
+struct infer_int_range_op_binder {
+  IntegerValueRange *bind_value;
+
+  infer_int_range_op_binder(IntegerValueRange *bind_value)
```
```suggestion
  explicit infer_int_range_op_binder(IntegerValueRange *bind_value)
```
or should this really be implicitly constructible?

gysit replied:
Good point!

Dinistro reviewed (commented):
Added a small nit comment, but this will be something that people with more experience with matchers should review.

kuhar commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -100,6 +101,41 @@ struct constant_op_binder {
+      return false;
+
+    // Set the range of all integer operands to the maximal range.
+    SmallVector<IntegerValueRange> argRanges;
+    argRanges.reserve(op->getNumOperands());
+    for (Value operand : op->getOperands())
+      argRanges.emplace_back(IntegerValueRange::getMaxRange(operand));
```
nit: I think we should be able to replace this with `map_to_vector`

kuhar commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -385,6 +446,13 @@ inline detail::constant_int_predicate_matcher m_NonZero() {
+/// unsigned integer range that does not contain zero. Note that this matcher
+/// interprets the target value as an unsigned integer.
+inline detail::constant_int_range_predicate_matcher m_NonZeroU() {
+  return {[](const ConstantIntRanges &range) { return range.umin().ugt(0); }};
```
What kind of overhead should we expect here vs. a simple APInt comparison?

gysit committed:
Address review comment.

krzysz00 commented:
Since our motivating example is "number of threads", perhaps we want to restrict this check to operations with no inputs?

gysit commented:
> Since our motivating example is "number of threads", perhaps we want to restrict this check to operations with no inputs?

Our "number of threads" operation has an optional integer argument. So deciding purely based on the number of inputs is not sufficient for our use case.

One way to make the change more opt in could be that we either add a new interface (i.e. a StaticIntRangeInterface or similar) that implements a `getStaticResultBounds` function that returns the static upper and lower bounds for the result of an operation (not taking the operands into account). Potentially that function could also be added to the existing interface, providing a default implementation that returns std::nullopt instead of a constant range. 

An alternative way could be to add a speculatable attribute on the division operation which if set makes the division speculatable. That way downstream passes could use potentially more complex analysis to decide if a division is speculatable.

Let me know if you have any preference?

Dinistro commented:
> One way to make the change more opt in could be that we either add a new interface (i.e. a StaticIntRangeInterface or similar) that implements a `getStaticResultBounds` function that returns the static upper and lower bounds for the result of an operation (not taking the operands into account). Potentially that function could also be added to the existing interface, providing a default implementation that returns std::nullopt instead of a constant range.

Duplicating the interface logic for this one matcher seems overkill. In the end, we would want this to be implemented almost everywhere possible.

> An alternative way could be to add a speculatable attribute on the division operation which if set makes the division speculatable. That way downstream passes could use potentially more complex analysis to decide if a division is speculatable.

I don't think that this is safe in the general case. Transformations need to be aware of this and strip this, which is not desirable. 

> Let me know if you have any preference?

I think that the current implementation is acceptable. While there is a slight constant overhead, this will only be relevant for users of this matcher, i.e., the division operation.

gysit commented:
> Duplicating the interface logic for this one matcher seems overkill. 

I agree that there is value in reusing the interface. And yes the overheads should be minor and only for very specific ops (divisions), while the benefit of hoisting divisions can be significant.

@kuhar @krzysz00 are you fine with moving forward with the current implementation?

kuhar commented:
> @kuhar @krzysz00 are you fine with moving forward with the current implementation?

I am not opposed to it, but don't have enough data to approve. It's not clear to me what actual overhead to expect and how much benefit to expect compared to plain constant checks.

gysit commented:
> I am not opposed to it, but don't have enough data to approve. It's not clear to me what actual overhead to expect and how much benefit to expect compared to plain constant checks.

I think the only place where I could see significant overhead is if the `inferResultRange` of an operation does some non-trivial computation which if it is the case should be fixed either way.

@zero9178 do you have a preference for using InferIntResultRange vs implementing a new operation interface for this specific case.

krzysz00 reviewed (approved):
I think, given how LICM works / that we can't plumb in an IntRangeAnalysis (which is something that'd make sense) this makes sense, though I wonder if we should maybe do this for signed divisions.

gysit committed:
Add support for signed divisions.

kuhar reviewed (commented):
The only comment I have is that the names are very close to `m_NonZero` yet do something quite different under the hood. I think it would make sense to either make `m_NonZero` do `m_nonZeroS || m_nonZeroU` (if possible?) or rename the new functions to something more verbose

gysit committed:
Rename the new matchers.

gysit commented:
> The only comment I have is that the names are very close to m_NonZero yet do something quite different under the hood.

Makes sense! I did rename to differentiate better. I left `m_NonZero` unchanged since it has other uses and I would suggest to keep it around as an option.

zero9178 reviewed (approved):
I think the way the PR discussion has been going to be right thing to do short-term. 
The way the speculation interface is being used is the way it was intended AFAIR. The interface lookup is also constant time and something that should be sped up in MLIR in general rather than being a blocker here IMO. 
What interface to use for the range analysis seems to be an ongoing issue since we now have I think 3 interfaces that roughly try to do range analysis in different scenarios so adding a 4th is definitely undesireable IMO. But at the same time the choice of the interface here shouldn't be a blocker. 

LGTM with the new names

gysit committed:
Fix an issue with the signed division.

Additionally, move the test cases to the existing
loop invariant code motion tests.

gysit commented:
There was a logic bug in my previous implementation of `getSpeculatibility` for the signed division. Additionally, I moved the tests to the existing licm tests. @Dinistro can you do a last pass if I got it right this time :).

Dinistro reviewed (approved):
LGTM!

gysit committed:
Fix check count.
